Separate docker compose environments and added documentation

rporres opened the pull request:
This change adds creates docker compose files:

* One to deploy using our [Docker images](https://hub.docker.com/r/pageonex/pageonex/) and the production settings. This is the easiest way to get started using our images and can be used to test how a real deployment would look like
* One to develop, which adds the possibility to build the application. This is useful to do a quick development without having to setup the whole environment (especially painful due the old versions we use)

It also adds a bunch of documentation on how to deploy pageonex in different scenarios.

numeroteca commented on doc/docker/running-pageonex-locally-with-docker-compose.md:
````diff
@@ -0,0 +1,55 @@
+
+Logs are not shown in the screen in detached mode, you can inspect them using
+```
+docker-compose up -d
````
Should it be
```
docker-compose logs
```
?
```suggestion
docker-compose logs
```

numeroteca commented on doc/docker/running-pageonex-locally-with-docker-compose.md:
```diff
@@ -0,0 +1,55 @@
+* A pageonex server running [the lastest pageonex image](https://hub.docker.com/r/pageonex/pageonex)
+* All data stored in separate volumes that will remain even if you delete your application's containers
+
+Once everything is running correctly, you can access the application from your browser in http:/localhost/3000
```
In my case it works (in chrome and firefox in ubuntu) at http://0.0.0.0:3000/
```suggestion
Once everything is running correctly, you can access the application from your browser in http:/localhost/3000 or http://0.0.0.0:3000/
```

rporres replied:
Fixed in 0b363c8

numeroteca commented on doc/docker/development-with-docker-compose.md:
```diff
@@ -0,0 +1,44 @@
+# Developing pageonex with docker-compose
+
+If you need to do development in pageonex without the hassle of installind all the dependencies you can use our [docker-compose-build file](/docker-compose-build.yml) which build the application from scratch.
```
```suggestion
If you need to do development in pageonex without the hassle of installing all the dependencies you can use our [docker-compose-build file](/docker-compose-build.yml) which builds the application from scratch.
```

numeroteca commented on doc/docker/development-with-docker-compose.md:
````diff
@@ -0,0 +1,44 @@
+
+* In order to start the pageonex and mysql containers, run from the root of this repository:
+```
+docker-compose -f docker-compose-build.yml up -d --build
````
It throws this error when running it:
`ERROR: Service 'app_build' depends on service 'mysql' which is undefined.`

rporres replied:
Fixed in e599b6c

numeroteca replied:
After the fix it works.
One more thing: I got this message "WARNING: Found orphan containers (pageonex-20181210_mysql_1) for this project. If you removed or renamed this service in your compose file, you can run this command with the --remove-orphans flag to clean it up." and the following error "for mysql_build  Cannot start service mysql_build: driver failed programming external connectivity on endpoint pageonex-20181210_mysql_build_1 (...) Bind for 0.0.0.0:13306 failed: port is already allocated", as I had already create the container locally with docker compose.

I don't know if it is worth mentioning it, see my suggestion. 
Is there an option to run in the same repository two containers so no need to remove database?

The text to ad would be as follows:

If you have previous docker compose installation of pageonex you'll need to remove the containers by adding `--remove-orphans` to the command. You'll lose all the data they contain (users, areas, threads...). Run from the root of this repository:
`docker-compose -f docker-compose-build.yml up -d --build --remove-orphans`
```

rporres replied:
I'll see if there's a way for docker-compose to avoid recognising both containers as coming from the same project... There's also the possibility of changing the port where mysql listens in one of the files, which is quite straight-forward

numeroteca replied:
I don't know if related, but I get this error (I don't know if this use case is important to be solved at this moment. Wondering how many people with try to switch between this two ways of working):

1. I run `docker-compose -f docker-compose-build.yml up -d --build` and it works good (works with my locally modified files).
2. Then I run  `docker-compose up` and it works good. I access another instance.
3. Then I run `docker-compose -f docker-compose-build.yml up -d --build`and I get this error and it doesn't launch the app:
```
Building app_build
Traceback (most recent call last):
  File "site-packages/docker/utils/build.py", line 179, in create_archive
PermissionError: [Errno 13] Permission denied: '/home/numeroteca/sites/pageonex/app/assets/images/threads/1/2/970/results.zip'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "bin/docker-compose", line 6, in <module>
  File "compose/cli/main.py", line 71, in main
  File "compose/cli/main.py", line 127, in perform_command
  File "compose/cli/main.py", line 1052, in up
  File "compose/cli/main.py", line 1048, in up
  File "compose/project.py", line 466, in up
  File "compose/service.py", line 329, in ensure_image_exists
  File "compose/service.py", line 1047, in build
  File "site-packages/docker/api/build.py", line 154, in build
  File "site-packages/docker/utils/build.py", line 31, in tar
  File "site-packages/docker/utils/build.py", line 183, in create_archive
OSError: Can not read file in context: /home/numeroteca/sites/pageonex/app/assets/images/threads/1/2/970/results.zip
[1441] Failed to execute script docker-compose
```

rporres replied:
I fail to see now how the main docker-compose can be affecting locally the other one. I'm starting to think it may be a good idea to create a separate issue for the development one since it is not the main purpose of this issue... We can work there any issue that appears. What do you think, @numeroteca?

numeroteca commented on README.md:
```diff
@@ -11,7 +11,15 @@ More Info
+* [You can run pageonex locally using docker-compose and our generated images](docs/docker/running-pageonex-locally-with-docker-compose.md). This is the easiest way.
+* [You can run pageonex locally using docker-compose and building locally the docker images](doc/docker/development-with-docker-compose.md). This is recommended if you want to do some development.
+* [You can install pageonex locally compiling in your system all the needed files](doc/local-install.md). This option is more advanced and it is only recommended if you're doing heavy development.
+* [You can use our Docker images to deploy pageonex agaings your mysql database](doc/docker/running-pageonex-in-your-environment.md). This is an advanced setup that it is useful if you're planning to maintain your own pageonex production environment.
```
```suggestion
* [You can use our Docker images to deploy pageonex against your mysql database](doc/docker/running-pageonex-in-your-environment.md). This is an advanced setup that it is useful if you're planning to maintain your own pageonex production environment.
```

rporres commented on doc/docker/running-pageonex-locally-with-docker-compose.md:
````diff
@@ -0,0 +1,55 @@
+```
+docker-compose exec mysql mysql -u root -h localhost --password=root dc-prod
+```
+
````
Fixed in 1a522ea

rporres commented:
Addressed comments @numeroteca 

I've left individual commits for review convenience. I can rearrange commits once you're done reviewing to avoid keeping trivial changes.

numeroteca commented on doc/docker/development-with-docker-compose.md:
````diff
@@ -0,0 +1,44 @@
+
+* To get a the rails console (See [INSTALL.md](/doc/INSTALL.md) for instructions on creating an admin user)
+```
+docker-compose -f docker-compose-build.yml exe app_build rails console
````
```suggestion
docker-compose -f docker-compose-build.yml exec app_build rails console
```

numeroteca commented on doc/docker/development-with-docker-compose.md:
````diff
@@ -0,0 +1,44 @@
+docker-compose -f docker-compose-build.yml exe app_build rails console
+```
+
+* If you make changes in the app and you want to rebuild your container
````
While developing, do changes need to be committed in order to see them locally after rebuilding? 
To test this, I made a small edit, then ran the command below, but I don't see the changes. How should be the normal development flow for local testing?

rporres replied:
I think `1aa9de8` fixes it. Can you test again?

numeroteca replied:
it works! 💃 
I don't need to commit or rebuild with docker-compose to see the changes. Great to make developing easier..

numeroteca commented:
🎉 👏🏾 👏🏾

rporres commented:
@rahulbot Can you create an issue describing your setup? We can take it from there
